plugins/providers/virtualbox/cap/cleanup_disks.rb

```diff
@@ -25,26 +26,77 @@ def self.cleanup_disks(machine, defined_disks, disk_meta_file)
-              disk_info = machine.provider.driver.get_port_and_device(d["uuid"])
+          storage_controllers = machine.provider.driver.read_storage_controllers
+          if storage_controllers.size ==  1
+            primary_controller = storage_controllers.first
+          else
+            primary_controller = storage_controllers.detect { |c| c.storage_bus == "SATA" }
+            if primary_controller.nil?
+              raise Vagrant::Errors::VirtualBoxDisksControllerNotFound, storage_bus: "SATA"
+            end
+          end
 
-              machine.ui.warn("Disk '#{d["name"]}' no longer exists in Vagrant config. Removing and closing medium from guest...", prefix: true)
+          primary = primary_controller.attachments.detect { |a| a[:port] == "0" && a[:device] == "0" }
+          if primary.nil?
+            raise Vagrant::Errors::VirtualBoxDisksPrimaryNotFound
+          end
+          primary_uuid = primary[:uuid]
 
-              if disk_info.empty?
-                LOGGER.warn("Disk '#{d["name"]}' not attached to guest, but still exists.")
+          if disk_meta
+            disk_meta.each do |d|
+              dsk = defined_disks.select { |dk| dk.name == d["name"] }
+              if !dsk.empty? || d["uuid"] == primary_uuid
+                next
               else
-                machine.provider.driver.remove_disk(disk_info[:port], disk_info[:device])
+                LOGGER.warn("Found disk not in Vagrantfile config: '#{d["name"]}'. Removing disk from guest #{machine.name}")
+                machine.ui.warn("Disk '#{d["name"]}' no longer exists in Vagrant config. Removing and closing medium from guest...", prefix: true)
+
+                controller = storage_controllers.detect { |c| c.name == d["controller"] }
+                attachment = controller.attachments.detect { |a| a[:port] == d["port"] &&
+                                                                 a[:device] == d["device"] }
+
+                # Retry lookup by UUID
+                if attachment.nil?
+                  attachment = controller.attachments.detect { |a| a[:uuid] == d["uuid"] }
+                end
+
+                if attachment.nil?
+                  LOGGER.warn("Disk '#{d["name"]}' not attached to guest, but still exists.")
+                else
+                  machine.provider.driver.remove_disk(controller.name, attachment[:port], attachment[:device])
+                end
+
+                machine.provider.driver.close_medium(d["uuid"])
               end
+            end
+          end
+        end
+
+        # @param [Vagrant::Machine] machine
+        # @param [VagrantPlugins::Kernel_V2::VagrantConfigDisk] defined_dvds
+        # @param [Hash] dvd_meta - A hash of all the previously defined dvds from the last configure_disk action
+        def self.handle_cleanup_dvd(machine, defined_dvds, dvd_meta)
+          if dvd_meta
```

Comment:
I think this should be true here too. dvd_meta should default to an empty hash, so the `.each` enumerator will cover this case without the extra if-statement.

Reply:
Sounds good! I added these checks originally because I didn't want to have to mock out an empty dvd_meta hash for the disk tests, and vice versa. But adding the type check seems to have revealed some errors in the tests anyways, so I'm happy to continue down that road.